 Sources/Class-Skeleton.php | 10 +++++++---
 1 file changed, 7 insertions(+), 3 deletions(-)

diff --git a/Sources/Class-Skeleton.php b/Sources/Class-Skeleton.php
index bab6411..fb5c92b 100644
--- a/Sources/Class-Skeleton.php
+++ b/Sources/Class-Skeleton.php
@@ -96,13 +96,17 @@ final class weSkeleton
 
 	/**
 	 * This is where we render the HTML page!
+	 *
+	 * @param string $from A starting layer. Leave empty to render the entire skeleton, as expected.
 	 */
 	function render($from = null)
 	{
+		if (!empty($this->layers['html']) || !empty($this->layers['body']))
+			loadTemplate('index');
 		if ($this->id === 'main' && empty($this->layers['default']))
 			fatal_lang_error('default_layer_missing');
 		$here = $from ? $this->layers[$from] : reset($this->skeleton);
